### PR TITLE
feat(Vault): add yield fee

### PR DESCRIPTION
**Pull request by PierrickGT**

### CODE REVIEW

File: src/Vault.sol [hunk 1]
```diff
@@ -292,10 +357,24 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
+    uint256 _liquidableYield = _availableBalanceOf(_tokenOut);
+    require(_liquidableYield >= _amountOut, "Vault/amount-gt-available-yield");
 
     _prizePool.contributePrizeTokens(address(this), _amountIn);
+
+    if (_yieldFeeRecipient != address(0) && _yieldFeePercentage != 0) {
+      // We get the percentage of liquidated yield and mint the equivalent amount in fees.
+      _mint(
+        _yieldFeeRecipient,
+        _getYieldFeeAmount(
```
Comment (asselstine) [inline src/Vault.sol:371]: **Reducing Operations for Better Accuracy**

This logic includes a lot of math operations! Let's reduce the number of operations to improve the floating point accuracy. We know that the `amountOut` is **after** the fee is taken, because of the `availableBalanceOf` fxn. To calculate the fee that should be taken as part of this amount, we can calculate the total:

We know that:

```
amountOut = total * (1 - feePercentage)
```

So to calculate `total` we can go

```
total = amountOut / (1-feePercentage)
```

The fee is the total less the amountOut:

```
fee = total - amountOut
```

So finally we can calculate the fee like so:

```
fee = amountOut / (1-feePercentage) - amountOut
```

Once we bake in the `FEE_PRECISION`, we see the code looks like this:

```
uint256 fee = (_amountOut * FEE_PRECISION) / (FEE_PRECISION - _yieldFeePercentage) - _amountOut
```

This equation uses far fewer operations!

File: src/Vault.sol [hunk 2]
```diff
@@ -618,7 +777,32 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
-    super._transfer(_from, _to, _shares);
+    _twabController.twabTransfer(_from, _to, uint112(_shares));
+  }
+
+  /**
+   * @notice Set claimer address.
+   * @param claimer_ Address of the claimer
+   */
+  function _setClaimer(Claimer claimer_) internal {
+    _claimer = claimer_;
```
Comment (asselstine) [inline src/Vault.sol:788]: Why is this a function?

File: src/Vault.sol [hunk 3]
```diff
@@ -618,7 +777,32 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
+    require(yieldFeePercentage_ <= FEE_PRECISION, "Vault/yieldFeePercentage-gt-1e9");
+    _yieldFeePercentage = yieldFeePercentage_;
+  }
+
+  /**
+   * @notice Set yield fee recipient address.
+   * @param yieldFeeRecipient_ Address of the fee recipient
+   */
+  function _setYieldFeeRecipient(address yieldFeeRecipient_) internal {
+    _yieldFeeRecipient = yieldFeeRecipient_;
```
Comment (asselstine) [inline src/Vault.sol:806]: Do we need this function?

**Review by asselstine — CHANGES_REQUESTED** [review]
Solid tests Pierrick!  Tho make sure to run the whole suite before you push :)

TLDR; 

let's switch to a "pull" model for fees. it'll make liquidations cheaper, and afford us the opportunity to cover undercollaterlization

File: src/Vault.sol [hunk 4]
```diff
@@ -596,8 +755,8 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
    * @dev `_receiver` cannot be the zero address.
    */
   function _mint(address _receiver, uint256 _shares) internal virtual override {
-    _twabController.twabMint(_receiver, uint112(convertToAssets(_shares)));
-    super._mint(_receiver, _shares);
+    _assetSupplyBalance += convertToAssets(_shares);
+    _twabController.twabMint(_receiver, uint112(_shares));
```
Comment (asselstine) [inline src/Vault.sol:759]: Vaults still need to be ERC20 compatible; this means it must emit the Transfer event as it would normally.

Only the *storage* is in the TwabController; the events still must be emitted here.

File: test/unit/Vault/Vault.t.sol [hunk 5]
```diff
@@ -342,22 +338,57 @@ contract VaultTest is UnitBaseSetup {
...
 
     vm.stopPrank();
   }
 
-  /* ============ helpers ============ */
-  function _setLiquidationPair() internal returns (address) {
-    return vault.setLiquidationPair(LiquidationPair(address(liquidationPair)));
+  /* ============ testSetYieldFeePercentage ============ */
+  function testSetYieldFeePercentage() public {
+    vm.expectEmit();
```
[the remark this answers is not in the document]
Comment (PierrickGT) [reply test/unit/Vault/Vault.t.sol:355]: You need to update forge-std: https://twitter.com/PaulRBerg/status/1634118576763203584
Now `vm.expectEmit()` is equivalent to `vm.expectEmit(true, true, true, true)`

File: src/Vault.sol (hunk 3 above)
Comment (PierrickGT) [reply src/Vault.sol:806]: Yes cause we use it in the constructor and in the `setYieldFeeRecipient` function.
https://github.com/pooltogether/v5-prize-vault/blob/3e9268fc221769398ae97eb26e4b5239eb130fce/src/Vault.sol#L177
https://github.com/pooltogether/v5-prize-vault/blob/3e9268fc221769398ae97eb26e4b5239eb130fce/src/Vault.sol#L497

File: src/Vault.sol (hunk 4 above)
Comment (PierrickGT) [reply src/Vault.sol:759]: Fixed in this commit: https://github.com/pooltogether/v5-vault/pull/6/commits/7831c1c1f0436ea59a445a9eae484b1e1356a380

File: test/unit/Vault/Vault.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (PierrickGT) [reply test/unit/Vault/Vault.t.sol:355]: Fixed in the following commits:
- https://github.com/pooltogether/v5-vault/pull/6/commits/9ca0d26ae26d72d9cfc4251ce77b2c4368078dc1
- https://github.com/pooltogether/v5-vault/pull/6/commits/ed841e51f11845b7879963c92bf5052dee7153f6

File: src/Vault.sol [hunk 6]
```diff
@@ -292,10 +357,24 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
 
-    require(availableBalanceOf(_tokenOut) >= _amountOut, "Vault/amount-gt-available-yield");
+    uint256 _liquidableYield = _availableBalanceOf(_tokenOut);
+    require(_liquidableYield >= _amountOut, "Vault/amount-gt-available-yield");
 
     _prizePool.contributePrizeTokens(address(this), _amountIn);
+
+    if (_yieldFeeRecipient != address(0) && _yieldFeePercentage != 0) {
+      // We get the percentage of liquidated yield and mint the equivalent amount in fees.
+      _mint(
```
[the remark this answers is not in the document]
Comment (PierrickGT) [reply src/Vault.sol]: Fixed in this commit: https://github.com/pooltogether/v5-vault/pull/6/commits/ed841e51f11845b7879963c92bf5052dee7153f6

File: src/Vault.sol [hunk 7]
```diff
@@ -364,22 +384,35 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
-          (_amountOut * _yieldFeePercentage) / _liquidableYield
-        )
-      );
+    if (_yieldFeePercentage != 0) {
+      uint256 _yieldFeeShares = (_amountOut * FEE_PRECISION) /
+        (FEE_PRECISION - _yieldFeePercentage) -
+        _amountOut;
+
+      _yieldFeeBalances[_yieldFeeRecipient] += _yieldFeeShares;
+      _yieldFeeTotalSupply += _yieldFeeShares;
```
Comment (asselstine) [inline src/Vault.sol:393]: Let's just track a single balance eh? I don't think they'll be changing recipients so often that we need a mapping.

Having to track per-recipient fees adds more than 7k to the gas cost of liquidation

File: src/Vault.sol [hunk 8]
```diff
@@ -570,28 +621,28 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
   function _availableBalanceOf(address _token) internal view returns (uint256) {
```
Comment (asselstine) [inline src/Vault.sol:621]: I feel like we should rename this function `_liquidatableBalanceOf` because it's too generic otherwise.  It gets confusing!

We can tackle this later.  Just noting it

File: src/Vault.sol [hunk 9]
```diff
@@ -67,6 +67,24 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
    */
   event ClaimerSet(Claimer previousClaimer, Claimer newClaimer);
 
+  /**
+   * @notice Emitted when a new LiquidationPair has been set.
+   * @param previousLiquidationPair Address of the previous liquidationPair
+   * @param newLiquidationPair Address of the new liquidationPair
+   */
+  event LiquidationPairSet(
+    LiquidationPair previousLiquidationPair,
```
Comment (asselstine) [inline src/Vault.sol]: No need to emit the previous liquidation pair; that data is available in the previous event

**Review by asselstine — CHANGES_REQUESTED** [review]

File: src/Vault.sol (hunk 9 above)
Comment (PierrickGT) [reply src/Vault.sol]: Fixed: https://github.com/pooltogether/v5-vault/pull/6/commits/3888878112483a5f5053c7e5dd3ce7aeb60bcaa3

File: src/Vault.sol (hunk 7 above)
Comment (PierrickGT) [reply src/Vault.sol:385]: Done: https://github.com/pooltogether/v5-vault/pull/6/commits/65ec5dd4559e77c24978642cbdc874f1451e5295

File: src/Vault.sol (hunk 1 above)
Comment (PierrickGT) [reply src/Vault.sol]: Fixed in the following commits:
- https://github.com/pooltogether/v5-vault/pull/6/commits/9ca0d26ae26d72d9cfc4251ce77b2c4368078dc1
- https://github.com/pooltogether/v5-vault/pull/6/commits/ed841e51f11845b7879963c92bf5052dee7153f6

**Review by asselstine — APPROVED** [review]
